### ISSUE TITLE
clear separation for machine-to-machine requirements

### ISSUE THREAD

**Issue by elarlang** [OWASP/ASVS#1724]
I think it makes sense to point more clearly out requirements, which are meant for machine-to-machine communication.

We can see many dimensions here:
* inside machine vs via some service
* internal network vs external network

Is the application as "client of the service" or as "service provider"

With quick scroll-through, I picked requirements which in my opinion are machine-to-machine requirements. For sure, there are more - especially related requirements for those which are listed here.

The point or question - can we use some structure changes to combine them somehow or we need to find some other solution.

# V1 Architecture, Design and Threat Modeling

## V1.2 Authentication Architecture

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **1.2.2** | [MODIFIED] Verify that communications between application components, including APIs, middleware and data layers, are authenticated and use individual user accounts. ([C3](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 306 |

## V1.4 Access Control Architecture

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **1.4.6** | [ADDED] Verify that communications between application components, including APIs, middleware and data layers, are performed with the least necessary privileges. ([C3](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 272 |

## V1.9 Communications Architecture

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **1.9.1** | Verify the application encrypts communications between components, particularly when these components are in different containers, systems, sites, or cloud providers. ([C3](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 319 |
| **1.9.2** | Verify that application components verify the authenticity of each side in a communication link to prevent person-in-the-middle attacks. For example, application components should validate TLS certificates and chains. | | ✓ | ✓ | 295 |

Related issue: https://github.com/OWASP/ASVS/issues/1443

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **1.14.1** | Verify the segregation of components of differing trust levels through well-defined security controls, firewall rules, API gateways, reverse proxies, cloud-based security groups, or similar mechanisms. | | ✓ | ✓ | 923 |
| **1.14.7** | [MOVED FROM 1.2.1] Verify the use of unique or special low-privilege operating system accounts for all application components, services, and servers. ([C3](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 250 |

# V2 Authentication

## V2.10 Service Authentication

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **2.10.1** | Verify that intra-service secrets do not rely on unchanging credentials such as passwords, API keys or shared accounts with privileged access. | | ✓ | ✓ | 287 | |
| **2.10.2** | [GRAMMAR] Verify that if passwords are required for service authentication, the service account used is not a default credential (e.g. root/root or admin/admin are default in some services during installation). | | ✓ | ✓ | 255 | |
| **2.10.4** | [GRAMMAR] Verify passwords, integrations with databases and third-party systems, seeds and internal secrets, and API keys are managed securely and not included in the source code or stored within source code repositories. Such storage should resist offline attacks. The use of a secure software key store (L1), hardware TPM, or an HSM (L3) is recommended for password storage. | | ✓ | ✓ | 798 | |

For 2.10.1 there is opened issue: https://github.com/OWASP/ASVS/issues/1032

# V3 Session Management

## V3.5 Token-based Session Management

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.5.2** | [GRAMMAR] Verify that the application uses session tokens rather than static API secrets and keys, except with legacy implementations. | | ✓ | ✓ | 798 | |

For 3.5.2 - sounds like 2.10.1 and 2.10.4, use the same CWE as 2.10.4 and has issue opened: https://github.com/OWASP/ASVS/issues/1522

# V6.4 Secret Management

## V6.4 Secret Management

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **6.4.1** | [MODIFIED] Verify that a secrets management solution such as a key vault is used to securely create, store, control access to and destroy secrets such as service account or 3rd party application credentials. ([C8](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 798 |
| **6.4.2** | Verify that key material is not exposed to the application but instead uses an isolated security module like a vault for cryptographic operations. ([C8](https://owasp.org/www-project-proactive-controls/#div-numbering)) | | ✓ | ✓ | 320 |

# V9 Communication Encryption

## V9.2 General Service to Service Communication Security

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **9.2.2** | Verify that encrypted communications such as TLS is used for all inbound and outbound connections, including for management ports, monitoring, authentication, API, or web service calls, database, cloud, serverless, mainframe, external, and partner connections. The server must not fall back to insecure or unencrypted protocols. | | ✓ | ✓ | 319 |
| **9.2.3** | Verify that all encrypted connections to external systems that involve sensitive information or functions are authenticated. | | ✓ | ✓ | 287 |

## V9.3 HTTPS Communication between Internal Services

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **9.3.1** | [ADDED] Verify that TLS or another appropriate transport encryption mechanism used for all connectivity between internal, HTTP-based services, and does not fall back to insecure or unencrypted communications. | | ✓ | ✓ | 319 |
| **9.3.2** | [MODIFIED, MOVED FROM 9.2.1] Verify that TLS connections between internal services use trusted certificates. Where internally generated or self-signed certificates are used, the consuming service must be configured to only trust specific internal CAs and specific self-signed certificates. All others should be rejected. | | ✓ | ✓ | 295 |
| **9.3.3** | [ADDED] Verify that mutual TLS (mTLS) is used by services communicating internally within a system or "intra-service communications" to ensure all the involved parties at each end of a network connection are who they claim to be. | | | ✓ | 295 |

# V12 File and Resources

## V12.6 SSRF Protection

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **12.6.1** | [GRAMMAR] Verify that the web or application server is configured with an allow list of resources or systems to which the server can send requests or load data or files from. | ✓ | ✓ | ✓ | 918 |

12.6.1 - related issue https://github.com/OWASP/ASVS/issues/1324

**Comment by tghosth** [comment]
I disagree with combining them into a single chapter, I think we need to make them clearer and potentially have the relevant requirements in a chapter grouped into a particular section.

**Comment by tghosth** [comment]
I agree that one of the primary things to be clarified should be whether the requirement is aimed at the developer of the service being provided or the developer of the process consuming a service.

**Comment by elarlang** [comment]
> I disagree with combining them into a single chapter, I think we need to make them clearer and potentially have the relevant requirements in a chapter grouped into a particular section.

It was not my proposal to combine them to single chapter. There are not enough requirements for this.

> I agree that one of the primary things to be clarified should be whether the requirement is aimed at the developer of the service being provided or the developer of the process consuming a service.

Based on subcategory title, I can say, that "machine-to-machine" separation is done for subcategories:
* V2.10 Service Authentication
* V9.2 General Service to Service Communication Security
* V9.3 HTTPS Communication between Internal Services

Which leaves us requirements:
* 1.2.2
* 1.4.6 (#1352 / Q5)
* 1.9.1 and 1.9.2 (probably will be solved by #1443)
* 1.14.1 (mentioned in #1227)
* 1.14.7 (it's a bit too close to 1.4.6)
* 3.5.2 (related issue #1522)
* 6.4.1, 6.4.2
* 12.6.1 (related issue #1324)

**Comment by tghosth** [comment]
> Which leaves us requirements:
> 
> * 1.2.2
> * 1.4.6 ([Access Control requirements reorg #1352](https://github.com/OWASP/ASVS/issues/1352) / Q5)
> * 1.9.1 and 1.9.2 (probably will be solved by [1.9.1, 1.9.2 - check duplicates and/or merge to V9 requirements #1443](https://github.com/OWASP/ASVS/issues/1443))
> * 1.14.1 (mentioned in [1.1.2 Threat modeling should be rephrased to cover different levels #1227](https://github.com/OWASP/ASVS/issues/1227))
> * 1.14.7 (it's a bit too close to 1.4.6)
> * 6.4.1, 6.4.2

So @elarlang is the action here to reword these requirements to be clearer that we are talking about machine to machine?

**Comment by elarlang** [comment]
> So @elarlang is the action here to reword these requirements to be clearer that we are talking about machine to machine?

Yes. Reword or relocation, whatever makes sense.

**Comment by tghosth** [comment]
Do you think #1767 makes this clearer?
